tools/net-tools/netstat.8
.TH netstat 8 MiNT-Net
.SH NAME
netstat \- show active network connections
.SH SYNOPSIS
.B "netstat [-f domain]"
.SH DESCRIPTION
.I Netstat
is used to display infomation about active communication
endpoints in either the
.B unix
or
.B inet
communication domain or both of them.
If invoked without arguments it shows only
.B unix
domain sockets.
You can specify the domain to deal with with the -f option.
.sp
For the unix domain the output of netstat consists of one line for
each socket in the system,
which contains the following entries:
.TP 15
.SM "Proto"
The protocol used by the socket, this can be one of:
.RS 15
.TP 15
.SM "unix"
This is the one and only protocol for unix sockets.
.TP
.SM "unknown"
The protocol is unknown, should never happen.
.RE
.TP 15
.SM "Flags"
These are the flags set on the socket. The output for this entry
consists of a space-separated list of tree-letter flag values.
Possible flags are:
.RS 15
.TP 15
.SM "ACC"
The socket is listening for incoming connections.
This applies only to stream sockets.
.TP
.SM "RCV"
The socket is shut down for receiving after a call to
.IR shutdown (3),
ie the socket cannot receive any messages.
.TP
.SM "SND"
The socket is shut down for sending after a call to
.IR shutdown (3),
ie it is impossible to send messages on the socket.
.TP
.SM "MRK"
If out-of-band data is handled inline, "receiving is at mark", ie
all data before the out of band data has been read.
.RE
.TP 15
.SM "State"
This field is the internal state of the socket,
the following values are possible:
.RS 15
.TP 15
.SM "VIRGIN"
The socket has just been created,
but has not been assigned a communication domain and type.
.TP
.SM "UNCONNECTED"
The socket is not connected to a peer.
.TP
.SM "CONNECTING"
The socket is about to connect to another one.
.TP
.SM "CONNECTED"
The socket is connected to another one.
.TP
.SM "DISONNECTING"
The socket is about to finish the connection to its peer.
.TP
.SM "DISCONNECTED"
The socket has finished a connection and is again unconnected.
.TP
.SM "UNKNOWN"
This should never happen.
.RE
.TP 15
.SM "Type"
This is the communication style the socket uses.
Possible values are:
.RS 15
.TP 15
.SM "STREAM"
The socket uses connection-oriented byte stream communication,
just like pipes do.
.TP
.SM "DGRAM"
The socket uses connectionless, unrelyable,
datagram oriented communication.
.TP
.SM "RAW"
This is a raw socket.
.TP
.SM "RDM"
Just like
.SM "DGRAM\c"
, except that you can rely on the packets to be delivered successfully.
.TP
.SM "SEQPACKET"
This is like
.SM "RDM\c"
, except that the receiver is guaranteed to get the packets in the
same order the sender sends them.
.TP
.SM "UNKNOWN"
This type should never happen.
.RE
.TP 15
.SM "Recv-Q"
The size of the next unread message for datagram oriented sockets,
and the number of unread bytes for stream oriented ones.
.TP
.SM "Address"
The local address of the socket, if
.IR bind (3)
was done on it.
This is a filename for the unix domain.
.SH OPTIONS
.TP 15
-f domain
Display information for active sockets in domain
.IR domain ,
overriding the default domain.
.SH FILES
.TP 15
/dev/unix
Unix domain information
.SH BUGS
Only unix domain implemented.
.SH AUTHOR
Kay Roemer <roemer@hera.rbi.informatik.uni-frankfurt.de>


tools/net-tools/netstat.c
/*
 *	This is a simple netstat(8). Currently the following options are
 *	implemented:
 *
 *	-f [inet|unix]	-- show inet/unix sockets
 *	-a		-- show all sockets, even if they are listening
 *			   TCP sockets which are not shown by default
 *
 *	(w) 1993,1994, Kay Roemer.
 */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/un.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <arpa/inet.h>
#include <netdb.h>

#include "../../sys/mint/net.h"


#define UNIX_DEVICE	"/dev/unix"
#define INET_DEVICE	"/dev/inet"

struct unix_info
{
	short	proto;			/* protcol numer, always 0 */
	short	flags;			/* socket flags, SO_* */
	short	type;			/* socket type, SOCK_DGRAM or SOCK_STREAM */
	short	state;			/* socket state, SS_* */
	short	qlen;			/* bytes in read buffer */
	short	addrlen;		/* addrlen, 0 if no address */
	struct sockaddr_un addr;	/* addr, only meaningful if addrlen > 0 */
};

struct inet_info
{
	short	proto;			/* protocol, IPPROTO_* */
	long	sendq;			/* bytes in send queue */
	long	recvq;			/* bytes in recv queue */
	short	state;			/* state */
	struct sockaddr_in laddr;	/* local address */
	struct sockaddr_in faddr;	/* foreign address */
};


static short show_unix_opt = 0;
static short show_usage_opt = 0;
static short show_inet_opt = 1;
static short show_all_opt = 0;


/*
 *	UNIX domain stuff.
 */

static const char *
un_decode_state (short state)
{
	switch (state)
	{
		case SS_VIRGIN:
			return "VIRGIN";

		case SS_ISUNCONNECTED:
			return "UNCONNECTED";

		case SS_ISCONNECTING:
			return "CONNECTING";

		case SS_ISCONNECTED:
			return "CONNECTED";

		case SS_ISDISCONNECTING:
			return "DISCONNECTING";

		case SS_ISDISCONNECTED:
			return "DISCONNECTED";
	}

	return "UNKNOWN";
}

static const char *
un_decode_type (short type)
{
	switch (type)
	{
		case SOCK_DGRAM:
			return "DGRAM";

		case SOCK_STREAM:
			return "STREAM";

		case SOCK_RDM:
			return "RDM";

		case SOCK_SEQPACKET:
			return "SEQPACKET";

		case SOCK_RAW:
			return "RAW";
	}

	return "UNKNOWN";
}

static const char *
un_decode_flags (short flags)
{
	static char strflags[20];

	strflags[0] = '\0';

	if (flags & SO_ACCEPTCON)
		strcat (strflags, "ACC ");
	if (flags & SO_CANTSNDMORE)
		strcat (strflags, "SND ");
	if (flags & SO_CANTRCVMORE)
		strcat (strflags, "RCV ");
	if (flags & SO_RCVATMARK)
		strcat (strflags, "MRK ");

	return strflags;
}

static void
show_unix (void)
{
	struct unix_info info;
	const char *proto;
	int fd, r;

	fd = open (UNIX_DEVICE, O_RDONLY);
	if (fd < 0)
	{
		fprintf (stderr, "cannot open device %s: %s\n",
			UNIX_DEVICE, strerror (errno));
		return;
	}

	puts ("Unix domain communication endpoints:");
	puts ("Proto   Flags            Type      State         Recv-Q Address");
	for (;;)
	{
		r = read (fd, &info, sizeof (info));
		if (r < 0)
		{
			fprintf (stderr, "cannot read next entry from %s: %s\n",
						UNIX_DEVICE, strerror (errno));
			break;
		}

		if (!r || r != sizeof (info))
			break;

		switch (info.proto)
		{
		case 0:
			proto = "unix";
			break;

		default:
			proto = "unknown";
			break;
		}

		printf ("%-7s %-16s %-9s %-13s %6u %s\n",
			proto,
			un_decode_flags (info.flags),
			un_decode_type (info.type),
			un_decode_state (info.state),
			info.qlen,
			info.addrlen ? info.addr.sun_path : "");
	}

	close (fd);
}

/*
 * INET domain stuff.
 */

static const char *
in_decode_state (short state)
{
	switch (state)
	{
		case TCPS_CLOSED:
			return "CLOSED";

		case TCPS_LISTEN:
			return "LISTEN";

		case TCPS_SYNSENT:
			return "SYNSENT";

		case TCPS_SYNRCVD:
			return "SYNRCVD";

		case TCPS_ESTABLISHED:
			return "ESTABLISHED";

		case TCPS_FINWAIT1:
			return "FINWAIT1";

		case TCPS_FINWAIT2:
			return "FINWAIT2";

		case TCPS_CLOSEWAIT:
			return "CLOSEWAIT";

		case TCPS_LASTACK:
			return "LASTACK";

		case TCPS_CLOSING:
			return "CLOSING";

		case TCPS_TIMEWAIT:
			return "TIMEWAIT";
	}

	return "UNKNOWN";
}

static char *
in_decode_addr (struct sockaddr_in *in, const char *proto)
{
	char buf[40];
	struct servent *sent;

	if (in->sin_port == 0)
	{
		strcpy(buf, "*:*");
	} else
	{
		sent = getservbyport (in->sin_port, proto);
		if (sent)
		{
			sprintf (buf, "%s:%s",
				in->sin_addr.s_addr == INADDR_ANY
					? "*" : inet_ntoa (in->sin_addr),
				sent->s_name);
		}
		else
		{
			sprintf (buf, "%s:%d",
				in->sin_addr.s_addr == INADDR_ANY
					? "*" : inet_ntoa (in->sin_addr),
				in->sin_port);
		}
	}

	return strdup (buf);
}

static const char *
in_decode_proto (short proto)
{
	switch (proto)
	{
		case IPPROTO_UDP:
			return "UDP";

		case IPPROTO_TCP:
			return "TCP";
	}

	return "UNKNOWN";
}

static void
show_udp (void)
{
	struct inet_info info;
	char *laddr, *faddr;
	int fd, r;

	fd = open (INET_DEVICE, O_RDONLY);
	if (fd < 0)
	{
		fprintf (stderr, "cannot open device %s: %s\n",
			INET_DEVICE, strerror (errno));
		return;
	}

	puts ("Active UDP connections");
	puts ("Proto  Recv-Q Send-Q Local-Address        "
	      "Foreign-Address      State");
	for (;;)
	{
		r = read (fd, &info, sizeof (info));
		if (r < 0)
		{
			fprintf (stderr, "cannot read next entry from %s: %s",
						INET_DEVICE, strerror (errno));
			break;
		}

		if (!r || r != sizeof (info))
			break;
		
		if (info.proto != IPPROTO_UDP)
			continue;

		laddr = in_decode_addr (&info.laddr, "udp");
		faddr = in_decode_addr (&info.faddr, "udp");

		printf ("%-6s %6lu %6lu %-20s %-20s %-15s\n",
			in_decode_proto (info.proto),
			info.recvq, info.sendq,
			laddr, faddr,
			in_decode_state (info.state));

		free (laddr);
		free (faddr);
	}

	close (fd);
}

static void
show_tcp (void)
{
	struct inet_info info;
	char *laddr, *faddr;
	int fd, r;

	fd = open (INET_DEVICE, O_RDONLY);
	if (fd < 0)
	{
		fprintf (stderr, "cannot open device %s: %s\n",
			INET_DEVICE, strerror (errno));
		return;
	}

	puts ("Active TCP connections");
	puts ("Proto  Recv-Q Send-Q Local-Address        "
	      "Foreign-Address      State");
	for (;;)
	{
		r = read (fd, &info, sizeof (info));
		if (r < 0) {
			fprintf (stderr, "cannot read next entry from %s: %s",
						INET_DEVICE, strerror (errno));
			break;
		}

		if (!r || r != sizeof (info))
			break;

		if (info.proto != IPPROTO_TCP)
			continue;

		if (!show_all_opt && info.state == TCPS_LISTEN)
			continue;


		laddr = in_decode_addr (&info.laddr, "tcp");
		faddr = in_decode_addr (&info.faddr, "tcp");

		printf ("%-6s %6lu %6lu %-20s %-20s %-15s\n",
			in_decode_proto (info.proto),
			info.recvq, info.sendq,
			laddr, faddr,
			in_decode_state (info.state));

		free (laddr);
		free (faddr);
	}

	close (fd);
}

static void
usage (void)
{
	puts ("Usage: netstat [options]");
	puts ("Options:");
	puts ("\t [-a] [-f [unix|inet]]");
}

int
main (int argc, char *argv[])
{
	int c;

	while ((c = getopt (argc, argv, "af:")) != EOF)
	{
		switch (c)
		{
			case 'f':
				if (!strcmp (optarg, "unix"))
					show_unix_opt = 1;
				else if (!strcmp (optarg, "inet"))
					show_inet_opt = 1;
				else
					show_usage_opt = 1;
				break;

			case 'a':
				show_all_opt = 1;
				break;

			case '?':
				show_usage_opt = 1;
				break;
		}
	}

	if (show_usage_opt)
	{
		usage ();
		return 0;
	}

	if (show_inet_opt)
	{
		show_tcp ();
		show_udp ();
	}

	if (show_unix_opt)	
		show_unix ();

	return 0;
}
